presto-main/src/main/java/com/facebook/presto/operator/OperatorSharedMemory.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+
+public class OperatorSharedMemory
+{
+    private int reservations = 0;
+    private Integer partitionKey;
+    private Long inMemorySizeInBytes;
+    private TaskContext taskContext;
+
+    public synchronized void setReservation(int partitionKey, OperatorContext operatorContext, long inMemorySizeInBytes)
+    {
+        if (this.reservations == 0) {
+            this.partitionKey = partitionKey;
+            this.inMemorySizeInBytes = inMemorySizeInBytes;
+            this.taskContext = operatorContext.getDriverContext().getPipelineContext().getTaskContext();
+            operatorContext.setMemoryReservation(inMemorySizeInBytes);
```

Comment:
this doesn't set reservation to `inMemorySizeInBytes` but always increase memory reservation by `inMemorySizeInBytes`. This is not `set***` method but `update***` method. Generally speaking `set*` are easier to use.